sdk/iot/core/src/az_iot_telemetry.c

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include <stdint.h>
+
+#include "az_iot_hub_client.h"
+#include <az_contract_internal.h>
+#include <az_result.h>
+#include <az_span.h>
+
+static const uint8_t telemetry_prop_delim = '?';
+static const uint8_t telemetry_prop_separator = '&';
+static const uint8_t telemetry_null_terminator = 0;
+static const az_span telemetry_topic_prefix = AZ_SPAN_LITERAL_FROM_STR("devices/");
+static const az_span telemetry_topic_modules_mid = AZ_SPAN_LITERAL_FROM_STR("/modules/");
+static const az_span telemetry_topic_suffix = AZ_SPAN_LITERAL_FROM_STR("/messages/events/");
+
+az_result az_iot_hub_client_telemetry_publish_topic_get(
+    az_iot_hub_client const* client,
+    az_iot_hub_client_properties const* properties,
+    az_span mqtt_topic,
+    az_span* out_mqtt_topic)
+{
+  AZ_CONTRACT_ARG_NOT_NULL(client);
+  AZ_CONTRACT_ARG_VALID_SPAN(mqtt_topic);
+  AZ_CONTRACT_ARG_NOT_NULL(out_mqtt_topic);
+
+  az_result result;
+
+  int32_t required_size = 0;
+
+  // Required topic parts
+  required_size = az_span_length(telemetry_topic_prefix)
+      + az_span_length(client->_internal.device_id) + az_span_length(telemetry_topic_suffix)
+      + sizeof(telemetry_null_terminator);
+
+  // Optional parts
+  if (!az_span_is_empty(client->_internal.options.module_id))
+  {
+    required_size += az_span_length(telemetry_topic_modules_mid);
+    required_size += az_span_length(client->_internal.options.module_id);
+  }
+  if (properties != NULL)
+  {
+    required_size
+        += az_span_length(properties->_internal.properties) + sizeof(telemetry_prop_delim);
+  }
+  if (!az_span_is_empty(client->_internal.options.user_agent))
+  {
+    required_size += az_span_length(client->_internal.options.user_agent)
+        + sizeof(telemetry_prop_delim);
+  }
+
+  // Only build topic if the span has the capacity
+  if (az_span_capacity(mqtt_topic) < required_size)
+  {
+    result = AZ_ERROR_OUT_OF_MEMORY;
+  }
+  else
+  {
+    AZ_RETURN_IF_FAILED(az_span_append(mqtt_topic, telemetry_topic_prefix, out_mqtt_topic));
+    AZ_RETURN_IF_FAILED(
+        az_span_append(*out_mqtt_topic, client->_internal.device_id, out_mqtt_topic));
+
+    if (az_span_length(client->_internal.options.module_id) != 0)
+    {
+      AZ_RETURN_IF_FAILED(
+          az_span_append(*out_mqtt_topic, telemetry_topic_modules_mid, out_mqtt_topic));
+      AZ_RETURN_IF_FAILED(
+          az_span_append(*out_mqtt_topic, client->_internal.options.module_id, out_mqtt_topic));
+    }
+
+    AZ_RETURN_IF_FAILED(az_span_append(*out_mqtt_topic, telemetry_topic_suffix, out_mqtt_topic));
+
+    if (properties != NULL)
+    {
+      AZ_RETURN_IF_FAILED(az_span_append_uint8(*out_mqtt_topic, telemetry_prop_delim, out_mqtt_topic));
+      AZ_RETURN_IF_FAILED(
+          az_span_append(*out_mqtt_topic, properties->_internal.properties, out_mqtt_topic));
+    }
+    if (az_span_length(client->_internal.options.user_agent) != 0)
```

Comment:
why do you have extra line in some places, but here you don't. nit: I prefer a blank line between if statements. This way people don't quickly read it as an else block

Reply:
that's fine with me. Just modeled it to how it makes sense in my head. Will space the `if`'s